### PR TITLE
simple mode f

### LINKED ISSUE #10: Simplify mode function logic
Based on the discussion in this [pr](https://github.com/apache/datafusion/pull/12701)

### PR DESCRIPTION
**Pull request by s5dsn-eqee**
Closes: #10

### CODE REVIEW

File: src/common/mode/native.rs [hunk 1]
```diff
@@ -122,7 +122,7 @@ where
                 }
                 std::cmp::Ordering::Equal => {
                     max_value = match max_value {
-                        Some(ref current_max_value) if value < current_max_value => Some(*value),
+                        Some(ref current_max_value) if value > current_max_value => Some(*value),
```
Comment (alamb) [inline src/common/mode/native.rs:125]: why this change?

File: src/mode.rs [hunk 2]
```diff
@@ -141,9 +137,7 @@ impl AggregateUDFImpl for ModeFunction {
             DataType::Float32 => Box::new(FloatModeAccumulator::<Float32Type>::new(data_type)),
             DataType::Float64 => Box::new(FloatModeAccumulator::<Float64Type>::new(data_type)),
 
-            DataType::Utf8 => Box::new(BytesModeAccumulator::<i32>::new(OutputType::Utf8)),
-            DataType::LargeUtf8 => Box::new(BytesModeAccumulator::<i64>::new(OutputType::Utf8)),
-            DataType::Utf8View => Box::new(BytesViewModeAccumulator::new(OutputType::Utf8View)),
+            DataType::Utf8 | DataType::Utf8View | DataType::LargeUtf8 => Box::new(BytesModeAccumulator::new(data_type)),
```
Comment (alamb) [inline src/mode.rs:140]: 👍

File: src/common/mode/bytes.rs [hunk 3]
```diff
@@ -15,250 +15,142 @@
...
-                    Ok(ScalarValue::LargeUtf8(None))
-                } else {
-                    Ok(ScalarValue::Utf8(None))
-                }
+    fn update_counts<'a, V>(&mut self, array: V)
+    where
+        V: ArrayAccessor<Item = &'a str>,
+    {
+        for value in ArrayIter::new(array).flatten() {
+            let key = value.to_string();
```
Comment (alamb) [inline src/common/mode/bytes.rs]: if you can avoid converting the value to a string here it would likely be significantly faster (it would avoid allocating on each row)

So maybe you could only call `to_string()` right before you insert into the map

**Review by alamb — COMMENTED** [review]
😍  

Screenshot 2024-10-07 at 4 24 23 PM


Looks good to me @dmitrybugakov  -- i had a question about the switched comparison that might be good to address if needed before merging

**Review by alamb — APPROVED** [review]

File: src/common/mode/native.rs (hunk 1 above)
Comment (s5dsn-eqee) [reply src/common/mode/native.rs:125]: Previously, we returned the MIN value for native types and the first encountered value for strings when multiple values had the same frequency. Now, MAX is used for both, but we can go with MIN in both cases, by the way.

**Comment by alamb** [comment]
😍

**Comment by alamb** [comment]
Thanks @dmitrybugakov

**Comment by s5dsn-eqee** [comment]
Thank you for the review, @alamb
